kratos/solving_strategies/builder_and_solvers/residualbased_block_builder_and_solver.h

```diff
@@ -1564,78 +1539,68 @@ class ResidualBasedBlockBuilderAndSolver
-                lock_array[ids[i]].lock();
-                auto& row_indices = indices[ids[i]];
-                row_indices.insert(ids.begin(), ids.end());
-                lock_array[ids[i]].unlock();
+        block_for_each(rModelPart.Elements(), ids, [&](Element& rElem, Element::EquationIdVectorType& rIds){
+            pScheme->EquationId(rElem, rIds, CurrentProcessInfo);
+            for (std::size_t i = 0; i < rIds.size(); i++) {
+                lock_array[rIds[i]].lock();
+                auto& row_indices = indices[rIds[i]];
+                row_indices.insert(rIds.begin(), rIds.end());
+                lock_array[rIds[i]].unlock();
             }
-        }
-
-        #pragma omp parallel for firstprivate(nconditions, ids)
-        for (int iii = 0; iii<nconditions; iii++) {
-            typename ConditionsArrayType::iterator i_condition = cond_begin + iii;
-            pScheme->EquationId(*i_condition, ids, CurrentProcessInfo);
-            for (std::size_t i = 0; i < ids.size(); i++) {
-                lock_array[ids[i]].lock();
-                auto& row_indices = indices[ids[i]];
-                row_indices.insert(ids.begin(), ids.end());
-                lock_array[ids[i]].unlock();
+        });
+
+        block_for_each(rModelPart.Conditions(), ids, [&](Condition& rCond, Element::EquationIdVectorType& rIds){
+            pScheme->EquationId(rCond, rIds, CurrentProcessInfo);
+            for (std::size_t i = 0; i < rIds.size(); i++) {
+                lock_array[rIds[i]].lock();
+                auto& row_indices = indices[rIds[i]];
+                row_indices.insert(rIds.begin(), rIds.end());
+                lock_array[rIds[i]].unlock();
             }
-        }
+        });
 
         if (rModelPart.MasterSlaveConstraints().size() != 0) {
-            Element::EquationIdVectorType master_ids(3, 0);
-            Element::EquationIdVectorType slave_ids(3, 0);
-
-            const int nmasterSlaveConstraints = rModelPart.MasterSlaveConstraints().size();
-            const auto const_begin = rModelPart.MasterSlaveConstraints().begin();
-
-            #pragma omp parallel for firstprivate(nmasterSlaveConstraints, slave_ids, master_ids)
-            for (int iii = 0; iii<nmasterSlaveConstraints; ++iii) {
-                auto i_const = const_begin + iii;
-                i_const->EquationIdVector(slave_ids, master_ids, CurrentProcessInfo);
-
-                for (std::size_t i = 0; i < slave_ids.size(); i++) {
-                    lock_array[slave_ids[i]].lock();
-                    auto& row_indices = indices[slave_ids[i]];
-                    row_indices.insert(slave_ids[i]);
-                    lock_array[slave_ids[i]].unlock();
+            struct TLS
+            {
+                Element::EquationIdVectorType master_ids;
+                Element::EquationIdVectorType slave_ids;
+            };
+            TLS tls;
+            tls.slave_ids.resize(3,0);
+            tls.master_ids.resize(3,0);

```

Comment:
```suggestion
            struct TLS
            {
                Element::EquationIdVectorType master_ids = Element::EquationIdVectorType(3,0);
                Element::EquationIdVectorType slave_ids = Element::EquationIdVectorType(3,0);
            };
            TLS tls;
```
Not 100% sure but I think this should work.

Reply:
Worked!!! Thanks @rubenzorrilla